File: solutions/TypeScript/[0438_Medium] Find All Anagrams in a String/findAnagrams.ts
function findAnagrams(s: string, p: string): number[] {
  const charOffset = 'a'.charCodeAt(0);
  const result: number[] = [];

  const pCharacters: number[] = new Array(26).fill(0);
  const sCharacters: number[] = new Array(26).fill(0);

  // First window
  for (let i = 0; i < p.length; i++) {
    pCharacters[p.charCodeAt(i) - charOffset]++;
    sCharacters[s.charCodeAt(i) - charOffset]++;
  }

  for (let i = p.length - 1; i < s.length; i++) {
    if (i >= p.length) {
      sCharacters[s.charCodeAt(i) - charOffset]++;
      sCharacters[s.charCodeAt(i - p.length) - charOffset]--;
    }

    let isAnagrams = true;

    for (let j = 0; j < 26; j++) {
      if (pCharacters[j] !== sCharacters[j]) {
        isAnagrams = false;
        break;
      }
    }

    if (isAnagrams) {
      result.push(i - (p.length - 1));
    }
  }

  return result;
}

File: solutions/Java/[0438_Medium] Find All Anagrams in a String/findAnagrams.java
class Solution {
  public List<Integer> findAnagrams(String s, String p) {
    List<Integer> result = new ArrayList<>();

    int[] pCharacters = new int[26];
    int[] sCharacters = new int[26];

    // First window
    for (int i = 0; i < p.length(); i++) {
      pCharacters[p.charAt(i) - 'a']++;

      if (i < s.length()) {
        sCharacters[s.charAt(i) - 'a']++;
      }
    }

    for (int j = p.length() - 1; j < s.length(); j++) {
      if (j >= p.length()) {
        sCharacters[s.charAt(j) - 'a']++;
        sCharacters[s.charAt(j - p.length()) - 'a']--;
      }

      boolean isAnagrams = true;

      for (int k = 0; k < 26; k++) {
        if (pCharacters[k] != sCharacters[k]) {
          isAnagrams = false;
          break;
        }
      }

      if (isAnagrams) {
        result.add(j - (p.length() - 1));
      }
    }

    return result;
  }
}
